src/include/simeng/branchpredictors/BranchPredictor.hh

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <cstdint>
+#include <tuple>
+
+namespace simeng {
+
+/** The types of branches recognised. */
+enum class BranchType {
+  Conditional = 0,
+  LoopClosing,
+  Return,
+  SubroutineCall,
+  Unconditional,
+  Unknown
+};
+
+/** A branch result prediction for an instruction. */
+struct BranchPrediction {
+  /** Whether the branch will be taken. */
+  bool isTaken;
+
+  /** The branch instruction's target address. If `isTaken == false`, the value
+   * will be ignored. */
+  uint64_t target;
+
+  /** Check for equality of two branch predictions . */
+  bool operator==(const BranchPrediction& other) {
+    if ((isTaken == other.isTaken) && (target == other.target))
+      return true;
+    else
+      return false;
+  }
+
+  /** Check for inequality of two branch predictions . */
+  bool operator!=(const BranchPrediction& other) {
+    if ((isTaken != other.isTaken) || (target != other.target))
+      return true;
+    else
+      return false;
+  }
+};
+
+/** An abstract branch predictor interface. */
+class BranchPredictor {
+ public:
+  virtual ~BranchPredictor(){};
+
+  /** Generate a branch prediction for the supplied instruction address, a
+   * branch type, and a known branch offset; defaults to 0 meaning offset is not
+   * known. Returns a branch direction and branch target address. */
+  virtual BranchPrediction predict(uint64_t address, BranchType type,
```

Comment:
This could be a terrible idea so unresolving. https://softwareengineering.stackexchange.com/questions/355074/is-overriding-a-pure-virtual-function-with-default-arguments-good-or-bad

Reply:
Good find @dANW34V3R. So can we revert and have no default in virtual but defaults in override?

Reply:
Or should we go down the suggested route of having `predict` be a non-overridable function which has the default value and then have an overridable function which `predict` calls?

Reply:
Yeah, it is a weird one. With this set up, if the object you are calling on has parent type `BranchPredictor` then the default value is used. If it has the derived type e.g. `PerceptronPredictor` and no value is given then it will error.

There are currently 2 usages of predict outside of the tests (in the fetch unit), so does this even need a default value. 

I guess you could have: 

```
  virtual BranchPrediction predict(uint64_t address, BranchType type,
                                   int64_t knownOffset) = 0;

  BranchPrediction predictWrapper(uint64_t address, BranchType type,
                                  int64_t knownOffset = 0) {
    return predict(address, type, knownOffset);
  }
```

as you suggest. Which isn't too bad. Needs a good comment (and better name)

Reply:
Given the minimal usage of this function as Dan said above, I think we should just ommit a default value entirely for this function

Reply:
Given Jack came to the different conclusion, lets postpone this for the Thursday meeting when Jack gets back

Reply:
Happy with the solution of no defaults. Leaving for Jack to resolve